checkout

```diff
@@ -0,0 +1,155 @@
+    echo
+}
+# --
+
+while [ 0 -lt $# ] ; do
+    case "$1" in
+    -h) usage; exit 0;;
+    -t) force_tag="true";       shift;;
+    -b) force_branch="true";    shift;;
+    -n) update_branch="false";  shift;;
+    *) if [ -n "$name" ] ; then
+            echo
+            echo "ERROR: only ONE tag-or-branch can be specified."
+            echo
+            usage
+            exit 1
+       fi
+       name="$1";      shift;;
+    esac
+done
+
+if [ -z "$name" ] ; then
+    echo
+    echo "ERROR: <tag-or-branch> is mandatory."
+    echo
+    usage
+    exit 1
+fi
+
+if $force_branch && $force_tag; then
+    echo
+    echo "ERROR: '-t' and '-b' are mutually exclusive. Please make up your mind."
+    echo
+    usage
+    exit 1
+fi
+# --
+
+# make sure submodules are initialised
+git submodule init
+for dir in sdk_container/src/third_party/coreos-overlay \
+           sdk_container/src/third_party/portage-stable ; do
+    if [ ! -f "$dir"/.git ] ; then
+        git submodule update -N "$dir"
+    fi
+done
+
+function check_all() {
+    local gitcmd="$1"
+    local name="$2"
+
+    local scripts="$(git $gitcmd \
+                        | sed -e 's/^[[:space:]]*//' -e 's:remotes/[^/]\+/::' \
+                        | grep -m1 -E "^$name\$")"
+    local overlay="$(git -C sdk_container/src/third_party/coreos-overlay $gitcmd \
+                        | sed -e 's/^[[:space:]]*//' -e 's:remotes/[^/]\+/::' \
+                        | grep -m1 -E "^$name\$")"
+    local portage="$(git -C sdk_container/src/third_party/portage-stable $gitcmd \
+                        | sed -e 's/^[[:space:]]*//' -e 's:remotes/[^/]\+/::' \
+                        | grep -m1 -E "^$name\$")"
+
+    if [ -n "$scripts" -a -n "$overlay" -a -n "$portage" ] ; then
+        echo "$scripts"
+    fi
+}
+# --
+
+#
+# TAG
+#
+if ! $force_branch; then 
+    for dir in . \
+        sdk_container/src/third_party/coreos-overlay \
+        sdk_container/src/third_party/portage-stable ; do
+        git -C "$dir" fetch --tags
+    done
+    tag="$(check_all 'tag -l' "$name")"
+
+    if [ -n "$tag" ] ; then
+        echo
+        echo "Checking out TAG '$tag'"
+        echo "----------------------------------"
+        git checkout "$tag"
+        git submodule update
+        exit
+    fi
+fi
+
+echo "No tag by name '$name' in repo + submodules."
+
+if $force_tag; then
+    echo "Tag-only mode forced, exiting."
+    exit 1
+fi
+
+#
+# BRANCH
+#
+
+branch="$(check_all "branch -a -l" "$name")"
+if [ -z "$branch" ]; then
+    echo "No branch by name '$name' in repo + submodules."
+    exit 1
+fi
+
+echo
+echo "Checking out BRANCH '$branch'"
+echo "----------------------------------"
+for dir in . \
+    sdk_container/src/third_party/coreos-overlay \
+    sdk_container/src/third_party/portage-stable ; do
+
+    git -C "$dir" checkout "$branch"
+    if $update_branch; then
+        echo "Pulling / rebasing '$dir'"
+        git -C "$dir" pull --rebase
```

Comment:
Why would you rebase? I think `--ff-only` is the better behavior in general.

Reply:
In our case I would even say that `reset --hard` is the an appropriate mode for the manged repos, e.g., when I want to build a PR branch that gets force pushed often. Would be nice to offer this through a flag.

Reply:
I'm trying to be least intrusive to local work while offering the best possible (automated) alignment with upstream.
With `pull --rebase`, the branches get updated but local changes are preserved. This way, contributors can work on multiple changes to multiple branches locally, and can rebase their trees on upstream changes when necessary.

Reply:
The rebase can lead to surprising results (e.g., a feature branch was force-pushed to drop a commit this would be sneaked in again, or?), I think ff-only is safer and when it errors the user should decide what to do when the local state differs - and probably choose between rebase or reset. Automatic rebase is normally not needed as long as the user wouldn't commit directly on a `flatcar-xy` branch but a feature branch.

Reply:
Maybe we can default to `pull --ff-only` and offer `pull --rebase` behind a `-r` and `reset --hard` behind a `-f` flag or so.

Reply:
Makes sense!